### PR TITLE
[Preview] Evaluate Object properties with an assignment

### PR DESCRIPTION
**Pull request by atwalg2**
Issue: #5695

I got this error, when adding my case. Has to do with an Eslint Complexity rule.
image

### Summary of Changes

* add getAncestor(number) function to find an ancestor of the current path
* allow 'a' in > dispatch({ a }) to be previewed
* allow 'a' in > function evaluate(script, { a } = {a: 2}) to be previewed

### Test Plan
- [x] make fixtures/other tests 
- [x] address complexity rule

Updated
Hit 'stuff' while running debugger on the page below, and try out some break points.
https://flax-sort.glitch.me/ (Updated: 03/26/2018)

### Gif
preview-fix

### CODE REVIEW

**Comment by jasonLaster** [comment]
This is looking good, i think the `evaluate` function needs some edits so that you can call it and test it...

not sure how you pause in there and can see the variables because they should be out of scope:

A couple of other bugs that i see, but are out of scope:

1. hovering on frameId highlights the entire line
2. hovering on console and log shouldn't do anything

**Comment by jasonLaster** [comment]
This PR would benefit from a lot more jest fixtures

File: src/components/QuickOpenModal.js [hunk 1]
```diff
@@ -357,8 +357,20 @@ export class QuickOpenModal extends Component<Props, State> {
...
+    const { symbols, selectedSource } = this.props;
+
+    if ((this.isFunctionQuery() || this.isVariableQuery()) && selectedSource) {
+      if (!symbols || symbols.functions.length == 0) {
+        return (
+          <div className="loading-indicator">{L10N.getStr("loadingText")}</div>
+        );
+      }
+    }
+  };
```
Comment (jasonLaster) [inline src/components/QuickOpenModal.js]: hmm, extra stuff

File: src/workers/parser/getSymbols.js [hunk 2]
```diff
@@ -227,7 +227,22 @@ function extractSymbol(path: SimplePath, symbols) {
...
-    if (t.isProperty(path.parent)) {
+    function keyIsValue(parent) {
+      if (
+        parent.key.start == parent.value.start &&
+        parent.key.loc.identifierName === parent.value.loc.identifierName
+      ) {
+        return true;
+      }
+      return false;
+    }
```
Comment (jasonLaster) [inline src/workers/parser/getSymbols.js]: lets move this to helpers

File: src/workers/parser/getSymbols.js [hunk 3]
```diff
@@ -227,7 +227,22 @@ function extractSymbol(path: SimplePath, symbols) {
...
 
-    if (t.isProperty(path.parent)) {
+    function keyIsValue(parent) {
+      if (
+        parent.key.start == parent.value.start &&
+        parent.key.loc.identifierName === parent.value.loc.identifierName
+      ) {
+        return true;
+      }
+      return false;
```
Comment (jasonLaster) [inline src/workers/parser/getSymbols.js]: we can return 

```js
const {key, value} = parent
return key.start == value.start && 
  key.loc.identifierName === value.loc.identifierName
```

Comment (jasonLaster) [inline src/workers/parser/getSymbols.js]: lots of jest tests!

File: src/workers/parser/getSymbols.js [hunk 4]
```diff
@@ -199,7 +200,12 @@ function extractSymbol(path: SimplePath, symbols) {
       return;
     }
 
-    if (t.isProperty(path.parent)) {
+    if (
+      path.parent &&
+      t.isProperty(path.parent) &&
+      !t.isAssignmentPattern(path.getAncestor(3)) &&
+      !(t.isCallExpression(path.getAncestor(3)) && keyIsValue(path.parent))
```
Comment (AnshulMalik) [inline src/workers/parser/getSymbols.js]: Line 206 and 207, can you please explain what they are doing?

Comment (atwalg2) [reply src/workers/parser/getSymbols.js]: basically this case says that if the grandparent in the AST is either a Assignment pattern, or a Call expression with one variable in it.... like for example dispatch({ a })

File: src/workers/parser/utils/helpers.js [hunk 5]
```diff
@@ -33,6 +33,14 @@ export function isYieldExpression(path: SimplePath) {
   );
 }
 
+export function keyIsValue(parent: Node): boolean {
```
Comment (jasonLaster) [inline src/workers/parser/utils/helpers.js]: lets add a comment here for what this is saying

File: src/workers/parser/getSymbols.js (hunk 4 above)
Comment (jasonLaster) [inline src/workers/parser/getSymbols.js]: perhaps we want to move these expressions to variables with comments

```js
const inAssignment = t.isAssignmentPattern(path.getAncestor(3));
const isFunctionParamShorthand = isObjectShorthand(path.parent) && t.isCallExpression(path.getAncestor(3));
```

Comment (jasonLaster) [inline src/workers/parser/getSymbols.js]: What's the use case for inAssignment?

Comment (jasonLaster) [inline src/workers/parser/getSymbols.js]: maybe keyIsValue should be isObjectShorthand

File: src/workers/parser/tests/fixtures/callExpressions.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+
+function callExpress() {
+    const y = ({ params });
+    const b = 9;
+    const d = 10;
+    const foo = ({ b });
```
Comment (jasonLaster) [inline src/workers/parser/tests/fixtures/callExpressions.js]: the code doesnt need to run, so you can remove some stuff

this is sufficient

```js
const y = ({ params });
const foo = ({ b });
return dispatch({ d });
```

File: src/workers/parser/getSymbols.js (hunk 4 above)
Comment (atwalg2) [reply src/workers/parser/getSymbols.js]: I knew their had to be a better way of saying this haha.

Comment (atwalg2) [reply src/workers/parser/getSymbols.js]: So the use case for the assignment pattern would be any time you have { blah } = {}, I may have to look at this again to make sure that it is working how I originally envisioned it.

File: src/workers/parser/getSymbols.js [hunk 7]
```diff
@@ -199,7 +200,16 @@ function extractSymbol(path: SimplePath, symbols) {
       return;
     }
 
-    if (t.isProperty(path.parent)) {
+    const inAssignment = t.isAssignmentPattern(path.getAncestor(3));
+    // explain what this is
+    const isFunctionParamShorthand =
```
Comment (jasonLaster) [inline src/workers/parser/getSymbols.js]: Hey sunny, can you explain what each of these variables is checking for? It'd be nice to have a comment next to each of them that explains their purpose and perhaps an example of something they'd catch

**Comment by atwalg2** [comment]
Still working on this. Will be making updates today. 03/26/2018

**Comment by jasonLaster** [comment]
Great!

**Comment by jasonLaster** [comment]
When I ran this branch locally I saw this error

**Comment by jasonLaster** [comment]
here are some notes:

```diff
diff --git a/src/workers/parser/getSymbols.js b/src/workers/parser/getSymbols.js
index f5c9bb6..634c637 100644
--- a/src/workers/parser/getSymbols.js
+++ b/src/workers/parser/getSymbols.js
@@ -201,20 +201,17 @@ function extractSymbol(path: SimplePath, symbols) {
       return;
     }
 
-    // Check assignment pattern (destructuring objects) myFunc({c} = {c: 2})
+    // e.g. myFunc({c} = {c: 2})
     const inAssignment = t.isAssignmentPattern(path.getAncestor(3));
 
-    /*   Checks if parsed element is a object shorthand contained in a call 
-    expression. Ex. myFunc({ a }) is shorthand for myFunc({a: a}), 
-    where a is a real value that is being assigned to a key of the same name. */
+    // e.g myFunc({ a })
     const isFunctionParamShorthand =
       isObjectShorthand(path.parent) && t.isCallExpression(path.getAncestor(3));
 
-    if (
-      t.isProperty(path.parent) &&
-      !inAssignment &&
-      !isFunctionParamShorthand
-    ) {
+    // Exclude identifiers that are not part of assignments or function param shorthands
+
+    // (sunny, this logic also seems wrong... not sure why though...)
+    if (!(inAssignment || isFunctionParamShorthand)) {
       return;
     }
 
diff --git a/src/workers/parser/tests/fixtures/callExpressions.js b/src/workers/parser/tests/fixtures/callExpressions.js
index 6915df0..2d31b33 100644
--- a/src/workers/parser/tests/fixtures/callExpressions.js
+++ b/src/workers/parser/tests/fixtures/callExpressions.js
@@ -1,3 +1,7 @@
 const y = ({ params });
 const foo = ({ b });
-return dispatch({ d });
\ No newline at end of file
+
+// we want to get `x`
+const {x} = obj;
+
+return dispatch({ d });
diff --git a/src/workers/parser/tests/fixtures/destructuring.js b/src/workers/parser/tests/fixtures/destructuring.js
index d425a0e..dde169b 100644
--- a/src/workers/parser/tests/fixtures/destructuring.js
+++ b/src/workers/parser/tests/fixtures/destructuring.js
@@ -15,6 +15,7 @@ let { [key]: foo } = { z: "bar" };
 
 let [, prefName] = prefsBlueprint[accessorName];
 
-function assignPatternFunc({c} = {c: 2}, {frameId} = {}) {
+// frameId could be simple and we'd want it to work
+function assignPatternFunc({c} = {c: 2}, { frameId }) {
   const v = c;
 }
diff --git a/src/workers/parser/utils/helpers.js b/src/workers/parser/utils/helpers.js
index 59d98a9..2af6657 100644
--- a/src/workers/parser/utils/helpers.js
+++ b/src/workers/parser/utils/helpers.js
@@ -35,7 +35,9 @@ export function isYieldExpression(path: SimplePath) {
 
 export function isObjectShorthand(parent: Node): boolean {
   const { key, value } = parent;
+
   return (
+    t.isProperty(parent) &&
     key &&
     key.start == value.start &&
     key.loc.identifierName === value.loc.identifierName
```

**Comment by atwalg2** [comment]
nice thanks for the notes. I will look to simplify things tonight.

**Comment by jasonLaster** [comment]
Just changed the PR title to `[Preview] Evaluate Object properties with an assignment` and changed the issue to: #5695

File: src/workers/parser/tests/__snapshots__/getSymbols.spec.js.snap [hunk 8]
```diff
@@ -92,6 +92,39 @@ imports:
...
+
+
+identifiers:
+
+
+classes:
+
+
+imports:
+
```
Comment (jasonLaster) [inline src/workers/parser/tests/__snapshots__/getSymbols.spec.js.snap]: why is call expression empty?

File: src/workers/parser/tests/__snapshots__/getSymbols.spec.js.snap [hunk 9]
```diff
@@ -372,6 +405,13 @@ variables:
 [(14, 6), (14, 16)]   foo
 [(14, 23), (14, 31)]   z
 [(16, 7), (16, 15)]   prefName
+[(18, 27), (18, 39)]
+[(18, 41), (18, 52)]
+[(18, 54), (18, 69)]
+[(18, 28), (18, 29)]   c
+[(18, 34), (18, 38)]   c
+[(18, 43), (18, 50)]   frameId
+[(19, 10), (19, 15)]   v
```
Comment (jasonLaster) [inline src/workers/parser/tests/__snapshots__/getSymbols.spec.js.snap]: some of these vars are missing

```diff
+ function assignPatternFunc({c} = {c: 2}, { frameId }, [a, b] = [1, 2]) {
+     const v = c;
```

actually looks like a & b and maybe assignPatternFunc

File: src/workers/parser/getSymbols.js [hunk 10]
```diff
@@ -212,7 +213,17 @@ function extractSymbol(path: SimplePath, symbols) {
...
+    const inAssignment = t.isAssignmentPattern(path.getAncestor(3));
+
+    // e.g myFunc({ a })
+    const isFunctionParamShorthand =
+      isObjectShorthand(path.parent) && t.isCallExpression(path.getAncestor(3));
+
+    if (
+      t.isProperty(path.parent) &&
+      !(inAssignment || isFunctionParamShorthand)
+    ) {
```
Comment (bomsy) [inline src/workers/parser/getSymbols.js]: correct me if i'm wrong, but this does not seem to give us the correct expression to evaluate. the tooltip just shows `undefined`

Comment (bomsy) [reply src/workers/parser/getSymbols.js]: Scratch dat!

I notice we are missing some use cases e.g 
- line 24 `frameActor` in `{ frameActor: frameId }`
- line 28 `b ` in `const foo = ({ b })` etc

Comment (bomsy) [reply src/workers/parser/getSymbols.js]: Hey @atwalg2 it is the behaviour we want ... but it seems like it only works for particular cases. 

If you don't mind i'll make some updates.

Comment (atwalg2) [reply src/workers/parser/getSymbols.js]: I definitely do not mind. I could use some help.

File: src/workers/parser/tests/fixtures/callExpressions.js [hunk 11]
```diff
@@ -0,0 +1,7 @@
+const y = ({ params });
+const foo = ({ b });
+const bar = { x: 3, y: "434", z: true, a: null, s: c * 2, d : getValue(2) }
+const params = frameId ? { frameActor: frameId } : {}
```
Comment (jasonLaster) [inline src/workers/parser/tests/fixtures/callExpressions.js]: these might belong in an assignment fixture

File: src/workers/parser/getSymbols.js [hunk 12]
```diff
@@ -279,6 +285,34 @@ function extractSymbol(path: SimplePath, symbols) {
...
+    }
+
+    if (t.isBooleanLiteral(value)) {
+      return `${value.value}`;
+    }
+
+    if (t.isCallExpression(value)) {
+      return "";
+    }
+  }
```
Comment (jasonLaster) [inline src/workers/parser/getSymbols.js]: maybe this should move to the helpers...

**Comment by jasonLaster** [comment]
just some flow stuff. Getting close!

File: src/workers/parser/tests/__snapshots__/getSymbols.spec.js.snap [hunk 13]
```diff
@@ -60,9 +60,13 @@ identifiers:
 [(8, 26), (8, 27)] b b
 [(10, 6), (18, 1)] Obj Obj
 [(10, 6), (10, 9)] Obj Obj
+[(11, 2), (11, 5)] 1 foo
 [(12, 2), (12, 9)] doThing doThing
 [(13, 4), (13, 11)] console console
 [(13, 12), (13, 15)] log log
+[(15, 2), (15, 14)] function () {
+  return 42;
+} doOtherThing
```
Comment (jasonLaster) [inline src/workers/parser/tests/__snapshots__/getSymbols.spec.js.snap]: hmm, this seems like a bit much... do we want this? maybe it is fine, just checking...

File: src/workers/parser/tests/__snapshots__/getSymbols.spec.js.snap [hunk 14]
```diff
@@ -92,6 +96,63 @@ imports:
 literals:
 
 
+hasJsx: false
+
+hasTypes: false"
```
Comment (jasonLaster) [inline src/workers/parser/tests/__snapshots__/getSymbols.spec.js.snap]: this too... what's this?

File: src/workers/parser/tests/__snapshots__/getSymbols.spec.js.snap [hunk 15]
```diff
@@ -286,20 +350,36 @@ identifiers:
 [(26, 17), (26, 25)] Backbone Backbone
 [(26, 26), (26, 30)] View View
 [(26, 31), (26, 37)] extend extend
+[(27, 2), (27, 9)] \\"li\\" tagName
+[(29, 2), (29, 8)] function () {
+  console.log(\\"yo\\");
+} render
```
Comment (jasonLaster) [inline src/workers/parser/tests/__snapshots__/getSymbols.spec.js.snap]: ...

**Pull request merged**